File: algorithms/cpp/652.cpp
#include <iostream>
#include <vector>
#include <unordered_map>
using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Solution {
private:
    unordered_map<string, vector<TreeNode *>> m;
    string serialize(TreeNode *node) {
        if (node == NULL) return "";
        string s = "(" + serialize(node->left) + to_string(node->val) + serialize(node->right) + ")";
        m[s].push_back(node);
        return s;
    }
public:
    vector<TreeNode *> findDuplicateSubtrees(TreeNode *root) {
        m.clear();
        serialize(root);
        vector<TreeNode *> result;
        for (auto &p: m) {
            if (p.second.size() > 1)
                result.push_back(p.second[0]);
        }
        return result;
    }
};

int main() {
    return 0;
}

File: algorithms/python3/652.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None
from collections import defaultdict

class Solution(object):
    def findDuplicateSubtrees(self, root):
        """
        :type root: TreeNode
        :rtype: List[TreeNode]
        """
        if not root: return []
        count = defaultdict(int)
        nodeDict = {}
        def helper(node):
            left = helper(node.left) if node.left else None
            right = helper(node.right) if node.right else None
            cur = (left, (node.val, ), right)
            count[cur] += 1
            nodeDict[cur] = node
            return cur
        helper(root)

        rls = []
        for k, v in count.items():
            if v > 1:
                rls.append(nodeDict[k])
        return rls
